### PR TITLE
Feature/add download links

### LINKED ISSUE #44: add download links to datasets
add an optional attribute to a dataset description doc, ie `dataset.json` (not a megaset) called `downloadLinks` that is an array of objects. Each object has required attibutes:
    {
        date: string,
        title: string,
        link: string
    },
### Update to dataset processing:
 On upload this data should be included in the `dataset- description` information. The schema setup should do this automatically.

### For updating the schema:

1. add a new file in `array-items` called "dataset-link.schema.json" that has the definition of one single dataset link, ie 
```date: string,
        title: string,
        link: string
```
2. reference that individual schema in `definitions.schema.json`, where 
```
datasetLinks:
    type: array
    items: 
        $ref: dataset-link.schema.json
        type: object 
```
3. update `input-dataset-info.schema.json` and `dataset-card.schema.json` to have datasetLinks that references the object def in `definitions.schema.json`

### Amendments to the datasets themselves: 
Add this data to the existing datasets (@meganrm will update this with which link goes to which dataset):
[
    {
        date: "May/26/2021",
        title: "The hiPSC single-cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Jul/21/2022",
        title: "The hiPSC single i1 cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_i1_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Jul/20/2022",
        title: "The hiPSC single M1 cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_m1_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Jul/21/2022",
        title: "The hiPSC single i2 cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_i2_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Jul/20/2022",
        title: "The hiPSC single M2 cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_m2_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Jul/19/2022",
        title: "The hiPSC single edge cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_edge_cell_image_dataset/tree/latest/metadata.csv",
    },
    {
        date: "Sep/28/2022",
        title: "The hiPSC single non-edge cell image dataset",
        link: "https://open.quiltdata.com/b/allencell/packages/aics/hipsc_single_nonedge_cell_image_dataset/tree/latest/metadata.csv",
    },

### Re-publish docs with new schema:
`npm run publish-docs`

### PR DESCRIPTION
**Pull request by rugeli**
Problem
=======
What is the problem this work solves, including
closes #44 

Solution
========
What I/we did to solve this problem
- created a new file called `dataset-link.schema.json` in `array-items`
- referenced the individual schema in `definition.schema.json`
- updated datasetLinks in other schema files
- import dataset-link schema in `full-schema.js` and `get-input-dataset-schema.js` in order to run `validate-datasets` and `process-dataset`
- amended dataset link data to each existing dataset   

with @meganrm 

## Type of change
Please delete options that are not relevant.
* New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: src/data-validation/schema/array-items/dataset-link.schema.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "$id": "array-items/dataset-link.schema.json",
+    "title": "Dataset Link",
+    "description": "Schema for each single dataset",
+    "additionalProperties": false,
+    "type": "object",
+    "properties": {
+        "date": {
+            "type": "string"
```
Comment (toloudis) [inline src/data-validation/schema/array-items/dataset-link.schema.json:10]: I can't remember what the date is supposed to mean here: is it a publication date or more like a file timestamp date?  I think you can put a description string here for documentation.

**Comment by toloudis** [comment]
Is there a follow-up modification to Cell Feature Explorer to then use this data?

File: src/data-validation/schema/array-items/dataset-link.schema.json (hunk 1 above)
Comment (rugeli) [reply src/data-validation/schema/array-items/dataset-link.schema.json:10]: I think a file timestamp data is associated with the key name - `dateCreated`, and the `date` here refers to the publication date. We can rename `date` to `dataReleased` to make it clear.

**Comment by rugeli** [comment]
> Is there a follow-up modification to Cell Feature Explorer to then use this data?

Good question, I ran `npm run publish-docs` to update the handoff specification docs after making all the changes in this pr, however I checked the docs there was no info about `downloadLinks` shown. I'll take a deeper look into it today.

File: src/process-single-dataset/index.js [hunk 2]
```diff
@@ -42,6 +42,18 @@ const processSingleDataset = async (
...
   const featureDefsData = await readFeatureData();
+  const featuresDataOrder = datasetJson.featuresDataOrder;
+  if (featuresDataOrder.length > featureDefsData.length) {
+    console.error("Error: Too many features in featuresDataOrder");
+    process.exit(1);
+  }
+  const keyList = Array.from(featureDefsData, (featureDataJson) => featureDataJson.key);
+  featuresDataOrder.forEach((keyName) => {
+    if (!keyList.includes(keyName)) {
+      console.error(`Error: key ${keyName} not found`);
```
Comment (toloudis) [inline src/process-single-dataset/index.js]: this should say `Error: key ${keyName} in featuresDataOrder does not exist in featureDefs`

File: src/process-single-dataset/index.js [hunk 3]
```diff
@@ -42,6 +42,18 @@ const processSingleDataset = async (
     datasetJson.featuresDataOrder.indexOf(defaultGroupBy);
 
   const featureDefsData = await readFeatureData();
+  const featuresDataOrder = datasetJson.featuresDataOrder;
+  if (featuresDataOrder.length > featureDefsData.length) {
+    console.error("Error: Too many features in featuresDataOrder");
```
Comment (toloudis) [inline src/process-single-dataset/index.js]: this should say `Error: featureDefs has ${featureDefsData.length} features but there are ${featuresDataOrder.length} listed in featuresDataOrder`

Comment (rugeli) [reply src/process-single-dataset/index.js]: Sure I'll make changes of them, thank you for the suggestions on error messages! I have created a new branch and moved the code over to it for the key validation issue.  

This PR should be ready to go if we don't have anymore changes on add links data.

**Review by toloudis — APPROVED** [review]

**Pull request merged**